File: fixtures/scale/dart/large/lib/ciphers/adfgvx/adfgvx_cipher.dart
import '../../core/cipher.dart';
import '../columnar/columnar_cipher.dart';
import '../columnar/columnar_key.dart';
import 'adfgvx_key.dart';
import 'adfgvx_symbols.dart';

/// Field cipher: substitution into ADFGVX symbols, then columnar transposition.
class AdfgvxCipher implements Cipher {
  final AdfgvxKey key;
  final ColumnarCipher _transposition;

  AdfgvxCipher(this.key) : _transposition = ColumnarCipher(ColumnarKey(key.transpositionWord));

  @override
  String name() => 'adfgvx';

  @override
  String encrypt(String plaintext) {
    final cleaned = plaintext.toUpperCase().replaceAll(RegExp(r'[^A-Z0-9]'), '');
    return _transposition.encrypt(substituteGrid(key.grid(), cleaned));
  }

  @override
  String decrypt(String ciphertext) =>
      unsubstituteGrid(key.grid(), _transposition.decrypt(ciphertext));
}

File: fixtures/scale/kotlin/large/src/cryptobench/ciphers/adfgvx/AdfgvxCipher.kt
package cryptobench.ciphers.adfgvx

import cryptobench.ciphers.columnar.ColumnarCipher
import cryptobench.ciphers.columnar.ColumnarKey
import cryptobench.core.Cipher

/** Field cipher: substitution into ADFGVX symbols, then columnar transposition. */
class AdfgvxCipher(private val key: AdfgvxKey) : Cipher {
    private val transposition = ColumnarCipher(ColumnarKey(key.transpositionWord))

    override fun name(): String = "adfgvx"

    override fun encrypt(plaintext: String): String {
        val cleaned = plaintext.uppercase().replace(Regex("[^A-Z0-9]"), "")
        return transposition.encrypt(AdfgvxSymbols.substitute(key.grid(), cleaned))
    }

    override fun decrypt(ciphertext: String): String =
        AdfgvxSymbols.unsubstitute(key.grid(), transposition.decrypt(ciphertext))
}
